### PR TITLE
Combine TensorExpression functions into one with if constexpr

### PR DESCRIPTION
**Pull request by macedo22**
## Proposed changes

`TensorExpression` currently defines two templated `get` functions and and two `operator~` overloads: one for the case when the `Derived` class is a `Tensor` and one for the case when it is a `TensorExpression`. This PR combines both of these `get` functions into one using `if constexpr`, and does the same for the two `operator~` overloads.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp [hunk 1]
```diff
@@ -307,33 +307,27 @@ struct TensorExpression<Derived, DataType, Symm, IndexList, ArgsList<Args...>> {
...
+  ///
   /// \returns const TensorExpression<Derived, DataType, Symm, IndexList,
   /// ArgsList<Args...>>&
-  template <typename V = Derived,
-            Requires<tt::is_a<Tensor, V>::value> = nullptr>
-  SPECTRE_ALWAYS_INLINE auto operator~() const {
-    return static_cast<const TensorExpression<Derived, DataType, Symm,
-                                              IndexList, ArgsList<Args...>>&>(
-        *this);
+  SPECTRE_ALWAYS_INLINE const auto& operator~() const {
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: can you mark this function `noexcept`? I think all of tensor expressions were written before we started marking things `noexcept`. So`() const {` -> `() const noexcept {`

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp [hunk 2]
```diff
@@ -307,33 +307,27 @@ struct TensorExpression<Derived, DataType, Symm, IndexList, ArgsList<Args...>> {
...
+    if constexpr (tt::is_a<Tensor, Derived>::value) {
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: We have a shortcut for the `::value`, so you can just use `tt::is_a_v<Tensor, Derived>`

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp [hunk 3]
```diff
@@ -405,50 +400,32 @@ struct TensorExpression<Derived, DataType, Symm, IndexList, ArgsList<Args...>> {
...
-  /// \tparam V used for SFINAE
   /// \param tensor_index the tensor component to retrieve
   /// \return the value of the DataType of component `tensor_index`
-  template <typename... LhsIndices, typename ArrayValueType,
-            typename V = Derived,
-            Requires<not tt::is_a<Tensor, V>::value> = nullptr>
+  template <typename... LhsIndices, typename ArrayValueType>
   SPECTRE_ALWAYS_INLINE type
   get(const std::array<ArrayValueType, num_tensor_indices>& tensor_index)
       const {
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: Also `noexcept`

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp [hunk 4]
```diff
@@ -405,50 +400,32 @@ struct TensorExpression<Derived, DataType, Symm, IndexList, ArgsList<Args...>> {
...
+      using rhs_only_with_lhs = rhs_elements_in_lhs<lhs, rhs>;
+      using transformation =
+          generate_transformation<rhs, lhs, rhs_only_with_lhs>;
+      return t_->get(
+          ComputeCorrectTensorIndex<transformation>::apply(tensor_index));
+    } else {
+      ASSERT(t_ == nullptr,
+             "A TensorExpression that shouldn't be holding a pointer to a "
+             "Tensor is holding one.");
+      return static_cast<const Derived&>(*this).template get<LhsIndices...>(
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: Can `static_cast<...>(*this)` be written as `(~*this)`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash the changes in right away :) Thanks for doing this refactor!

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp (hunk 1 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: done!

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp (hunk 2 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: done!

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp (hunk 3 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: done!

File: src/DataStructures/Tensor/Expressions/TensorExpression.hpp (hunk 4 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/TensorExpression.hpp]: done!

**Review by nilsdeppe — APPROVED** [review]
Looks good! :+1:

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**